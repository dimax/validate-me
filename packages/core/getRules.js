export const datePatterns = {
  date: /^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])$/,
  time: /^([01]\d|2[0-3]):[0-5]\d$/,
  'datetime-local': /^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])T([01]\d|2[0-3]):[0-5]\d$/,
  week: /^\d{4}-W([0-4]\d|5[0-3])$/,
  month: /^\d{4}-(0[1-9]|1[0-2])$/,
};
const dateExamples = {
  date: '2019-03-24',
  time: '14:59',
  'datetime-local': '2019-02-28T23:59',
  week: '2019-W53',
  month: '2019-12',
};
const number = 'number';
const color = 'color';
const range = 'range';
const min = 'min';
const max = 'max';
const patternName = 'pattern';

/**
 *
 * @param {string} type Type
 * @param {any} props Props
 * @return {Array<Array<string>>} Configured rules
 */
export default function getRules(type, props) {
  const rules = props.required ? [['required']] : [];
  const addRule = rules.push.bind(rules);

  if (type === color) {
    addRule([color]);
  } else if (type === number || type === range) {
    addRule([number]);
    props.min && addRule([min, props.min]);
    props.max && addRule([max, props.max]);
  } else if (type in dateExamples) {
    const pattern = datePatterns[type];

    addRule([patternName, pattern, '', dateExamples[type]]);

    if (props.min) {
      if (process.env.NODE_ENV !== 'production' && !pattern.test(props.min)) {
        throw new Error(
          `[dev-only] @validate-me: the value "${
            props.min
          }" from the min prop must be a valid ${type} value.`,
        );
      }
      addRule([min, props.min]);
    }
    if (props.max) {
      if (process.env.NODE_ENV !== 'production' && !pattern.test(props.max)) {
        throw new Error(
          `[dev-only] @validate-me: the value "${
            props.max
          }" from the max prop must be a valid ${type} value.`,
        );
      }
      addRule([max, props.max]);
    }
  }
  if (props.pattern) {
    const rule = [patternName, props.pattern];

    if (type === 'email' && props.multiple) {
      rule.push('mul');
    }

    addRule(rule);
  }

  return rules;
}
