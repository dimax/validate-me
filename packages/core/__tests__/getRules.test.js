import getRules from '../getRules';

describe('core/getRules', () => {
  test('Build rules for color inputs', () => {
    expect(getRules('color', {})).toMatchSnapshot();
  });

  test('Build rules for number inputs', () => {
    expect(
      getRules('number', {
        required: true,
        min: '0',
        max: '2',
      }),
    ).toMatchSnapshot();
    expect(
      getRules('range', {
        min: '0',
        max: '2',
      }),
    ).toMatchSnapshot();
  });

  test('Build rules for dates inputs with valid props', () => {
    expect(
      getRules('date', {
        min: '2018-01-03',
        max: '2018-05-03',
      }),
    ).toMatchSnapshot();
    expect(
      getRules('time', {
        min: '14:59',
        max: '23:00',
      }),
    ).toMatchSnapshot();
    expect(
      getRules('datetime-local', {
        max: '2019-02-28T23:59',
      }),
    ).toMatchSnapshot();
    expect(
      getRules('week', {
        min: '2019-W01',
        max: '2019-W53',
      }),
    ).toMatchSnapshot();
    expect(
      getRules('month', {
        min: '2019-01',
      }),
    ).toMatchSnapshot();
  });

  test('Fail to build rules for dates inputs with invalid props', () => {
    expect(() =>
      getRules('date', {
        max: '2018-2-03',
      }),
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      getRules('time', {
        min: '24:00',
      }),
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      getRules('datetime-local', {
        max: '2019-02-28T24:00',
      }),
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      getRules('week', {
        min: '2019-W65',
      }),
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      getRules('month', {
        max: '2019-15',
      }),
    ).toThrowErrorMatchingSnapshot();
  });

  test('Build rules for email inputs', () => {
    expect(
      getRules('email', {
        multiple: true,
      }),
    ).toMatchSnapshot();
    expect(
      getRules('email', {
        pattern: '.+@gmail.com',
      }),
    ).toMatchSnapshot();
  });

  test('Build rules for text inputs', () => {
    expect(
      getRules('text', {
        pattern: '\\w+',
      }),
    ).toMatchSnapshot();
  });

  test('Build rules for select inputs', () => {
    expect(getRules('select', {})).toMatchSnapshot();
  });
});
