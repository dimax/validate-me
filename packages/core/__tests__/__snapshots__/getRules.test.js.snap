// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/getRules Build rules for color inputs 1`] = `
Array [
  Array [
    "color",
  ],
]
`;

exports[`core/getRules Build rules for dates inputs with valid props 1`] = `
Array [
  Array [
    "pattern",
    /\\^\\\\d\\{4\\}-\\(0\\[1-9\\]\\|1\\[0-2\\]\\)-\\(0\\[1-9\\]\\|\\[12\\]\\\\d\\|3\\[01\\]\\)\\$/,
    "",
    "",
    "2019-03-24",
  ],
  Array [
    "min",
    "2018-01-03",
  ],
  Array [
    "max",
    "2018-05-03",
  ],
]
`;

exports[`core/getRules Build rules for dates inputs with valid props 2`] = `
Array [
  Array [
    "pattern",
    /\\^\\(\\[01\\]\\\\d\\|2\\[0-3\\]\\):\\[0-5\\]\\\\d\\$/,
    "",
    "",
    "14:59",
  ],
  Array [
    "min",
    "14:59",
  ],
  Array [
    "max",
    "23:00",
  ],
]
`;

exports[`core/getRules Build rules for dates inputs with valid props 3`] = `
Array [
  Array [
    "pattern",
    /\\^\\\\d\\{4\\}-\\(0\\[1-9\\]\\|1\\[0-2\\]\\)-\\(0\\[1-9\\]\\|\\[12\\]\\\\d\\|3\\[01\\]\\)T\\(\\[01\\]\\\\d\\|2\\[0-3\\]\\):\\[0-5\\]\\\\d\\$/,
    "",
    "",
    "2019-02-28T23:59",
  ],
  Array [
    "max",
    "2019-02-28T23:59",
  ],
]
`;

exports[`core/getRules Build rules for dates inputs with valid props 4`] = `
Array [
  Array [
    "pattern",
    /\\^\\\\d\\{4\\}-W\\(\\[0-4\\]\\\\d\\|5\\[0-3\\]\\)\\$/,
    "",
    "",
    "2019-W53",
  ],
  Array [
    "min",
    "2019-W01",
  ],
  Array [
    "max",
    "2019-W53",
  ],
]
`;

exports[`core/getRules Build rules for dates inputs with valid props 5`] = `
Array [
  Array [
    "pattern",
    /\\^\\\\d\\{4\\}-\\(0\\[1-9\\]\\|1\\[0-2\\]\\)\\$/,
    "",
    "",
    "2019-12",
  ],
  Array [
    "min",
    "2019-01",
  ],
]
`;

exports[`core/getRules Build rules for email inputs 1`] = `
Array [
  Array [
    "pattern",
    "[a-z0-9!#$%&'*+/=?^_\`{|}~.-]+@[a-z0-9-]+(.[a-z0-9-]+)*",
    "i",
    "mul",
  ],
]
`;

exports[`core/getRules Build rules for email inputs 2`] = `
Array [
  Array [
    "pattern",
    ".+@gmail.com",
    "i",
  ],
]
`;

exports[`core/getRules Build rules for number inputs 1`] = `
Array [
  Array [
    "required",
  ],
  Array [
    "number",
  ],
  Array [
    "min",
    "0",
  ],
  Array [
    "max",
    "2",
  ],
]
`;

exports[`core/getRules Build rules for number inputs 2`] = `
Array [
  Array [
    "number",
  ],
  Array [
    "min",
    "0",
  ],
  Array [
    "max",
    "2",
  ],
]
`;

exports[`core/getRules Build rules for select inputs 1`] = `Array []`;

exports[`core/getRules Build rules for text inputs 1`] = `
Array [
  Array [
    "pattern",
    "\\\\w+",
  ],
]
`;

exports[`core/getRules Fail to build rules for dates inputs with invalid props 1`] = `"[dev-only] @validate-me: the value \\"2018-2-03\\" from the max prop must be a valid date value."`;

exports[`core/getRules Fail to build rules for dates inputs with invalid props 2`] = `"[dev-only] @validate-me: the value \\"24:00\\" from the min prop must be a valid time value."`;

exports[`core/getRules Fail to build rules for dates inputs with invalid props 3`] = `"[dev-only] @validate-me: the value \\"2019-02-28T24:00\\" from the max prop must be a valid datetime-local value."`;

exports[`core/getRules Fail to build rules for dates inputs with invalid props 4`] = `"[dev-only] @validate-me: the value \\"2019-W65\\" from the min prop must be a valid week value."`;

exports[`core/getRules Fail to build rules for dates inputs with invalid props 5`] = `"[dev-only] @validate-me: the value \\"2019-15\\" from the max prop must be a valid month value."`;
